### PR TITLE
HHH-17817 - Add option to Enable Byte Code Enhancement for specific classes

### PR DESCRIPTION
**Pull request by Akshit97**
https://hibernate.atlassian.net/browse/HHH-17817

### CODE REVIEW

File: tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java [hunk 1]
```diff
@@ -168,6 +173,12 @@ public boolean doExtendedEnhancement(UnloadedClass classDescriptor) {
 			}
 			for ( File file : sourceSet ) {
 
+				String className = file.getName().replace(".class", "");
```
Comment (sebersole) [inline tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java:176]: I think this part is incorrect.  I think it needs to stip file paths up to the package name and convert filesystem path separators into `.`., right?

File: tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java [hunk 2]
```diff
@@ -38,6 +39,7 @@ public EnhancementSpec(HibernateOrmSpec ormDsl, Project project) {
 		enableDirtyTracking = makeProperty( project ).convention( true );
 		enableAssociationManagement = makeProperty( project ).convention( false );
 		enableExtendedEnhancement = makeProperty( project ).convention( false );
+		classNames = project.getObjects().property(String.class).convention("");
```
Comment (sebersole) [inline tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java:42]:
```suggestion
		classNames = project.getObjects().listProperty(String.class).convention("");
```

File: tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java [hunk 3]
```diff
@@ -195,4 +204,8 @@ public void enableExtendedEnhancement(boolean enable) {
 	public void extendedEnhancement(boolean enable) {
 		setEnableExtendedEnhancement( enable );
 	}
+
+	public void setClassNames(String classNames) {
+		this.classNames.set(classNames);
+	}
```
Comment (sebersole) [inline tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java:210]:
```suggestion

	public void setClassNames(List<String> classNames) {
		this.classNames.set(classNames);
	}
	
	public void setClassNames(Provider<List<String>> classNames) {
		this.classNames.set(classNames);
	}
	
	public void includeClassName(String className) {
		this.classNames.add(className);
	}

	...
```

**Review by sebersole — CHANGES_REQUESTED** [review]
Few comments

File: tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java (hunk 1 above)
Comment (Akshit97) [reply tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java:176]: No, `file.getName()` will return only the file name and not the complete path. So, just removing `.class` from it works.

File: tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java [hunk 4]
```diff
@@ -28,6 +28,7 @@ public class EnhancementSpec {
 	private final Property<Boolean> enableDirtyTracking;
 	private final Property<Boolean> enableAssociationManagement;
 	private final Property<Boolean> enableExtendedEnhancement;
+	private final Property<String> classNames;
```
[the remark this answers is not in the document]
Comment (Akshit97) [reply tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/enhance/EnhancementSpec.java]: Done

**Comment by Akshit97** [comment]
> 

@sebersole Updated the PR as per your comments.

File: tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java (hunk 1 above)
Comment (sebersole) [reply tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java:176]: Not normal, but what about same-named classes in different packages?

Comment (Akshit97) [reply tooling/hibernate-enhance-maven-plugin/src/main/java/org/hibernate/orm/tooling/maven/MavenEnhancePlugin.java]: Yes, in that case both will be enhanced as per current implementation.
Updated it to use the fully qualified class name using your implementation of `determineClassName`

**Comment by sebersole** [comment]
Ok, this looks great!

One last thing, could you squash these commits?  Thanks.

**Comment by Akshit97** [comment]
@sebersole Done

**Pull request merged**